### PR TITLE
MAINT: test-refactor of `numpy/_core/numeric.py`

### PR DESCRIPTION
**Pull request by Carreau**
This is a pure-test refactor (as well as adding a few cases), extracted from numpy/numpy#24663.

Testing each individually has a small Pytest overhead, but has the advantage of each test being able to fail independently in case of a modification of the tested function.

It also make it much easier to add a new test, and I believe is much more readable.

### CODE REVIEW

**Comment by ngoldbaum** [comment]
Test failure is unrelated

**Comment by Carreau** [comment]
Thanks @rgommers for the rebase

File: numpy/_core/tests/test_numeric.py [hunk 1]
```diff
@@ -1968,57 +1968,145 @@ def test_base_range(self):
...
+
+# Non-nanable – those cannot hold nans
+_a12 = np.array([1, 2])
+_a12b = _a12.copy()
+_a123 = np.array([1, 2, 3])
+_a13 = np.array([1, 3])
+_a34 = np.array([3, 4])
+
+
+# those can hold nan
```
Comment (seberg) [inline numpy/_core/tests/test_numeric.py:1989]:
```suggestion
```
I think further down?

File: numpy/_core/tests/test_numeric.py [hunk 2]
```diff
@@ -1968,57 +1968,145 @@ def test_base_range(self):
...
+_aS1 = np.array(["a"], dtype="S1")
+_aS1b = _aS1.copy()
+_aS1u4 = np.array([("a", 1)], dtype="S1,u4")
+_aS1u4b = _aS1u4.copy()
+
+
```
Comment (seberg) [inline numpy/_core/tests/test_numeric.py:1995]:
```suggestion
```

**Review by seberg — COMMENTED** [review]
Sorry to nitpick, I can also live with this.  But these globals seem a bit annoying.  Can we rewrite it as (can also be in the class even if the class is useless)
```
def _test_array_equal_parametrizations():
    # those are 0-d arrays, it used to be a special case
    # where (_e0 == _e0).all() would raise
    e0 = np.array(0, dtype='int')
    e1 = np.array(1, dtype='float')

    yield (e0, e0.copy(), None, True)
    yield ...
```
that would keep locality of the variables and comparisons and just remove the weird globals.

EDIT: Which of course means using:
```
@pytest.mark.parametrize("...", _test_array_equal_parametrizations())
```

**Review by seberg — APPROVED** [review]
Thanks a lot! Much easier to follow now with things being closer :).

**Pull request merged**